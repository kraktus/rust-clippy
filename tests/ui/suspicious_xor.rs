#![allow(unused)]
#![warn(clippy::suspicious_xor)]
#![allow(clippy::eq_op)]

macro_rules! macro_test {
    () => {
        13
    };
}

macro_rules! macro_test_inside {
    () => {
        1 ^ 2 // should warn even if inside macro
    };
}

fn main() {
    // Should warn:
    let _ = 2 ^ 5;
    let _ = 2i32 ^ 9i32;
    let _ = 2i32 ^ 2i32;
    let _ = 50i32 ^ 3i32;
    let _ = 5i32 ^ 8i32;
    let _ = 2i32 ^ 32i32;
    macro_test_inside!();

    // Should not warn:
    let _ = 0x02 ^ 5;
    let _ = 2 ^ 0x02;
    let _ = 10 ^ 0b0101;
    let _ = 2i32 ^ macro_test!();
}
