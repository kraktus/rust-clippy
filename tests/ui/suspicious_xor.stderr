error: '^' is not the exponentiation operator
  --> $DIR/suspicious_xor.rs:13:13
   |
LL |     let _ = 2 ^ 5;
   |             ^^^^^ help: did you mean to write: `2.pow(5)`
   |
   = note: `-D clippy::suspicious-xor` implied by `-D warnings`

error: '^' is not the exponentiation operator
  --> $DIR/suspicious_xor.rs:14:13
   |
LL |     let _ = 2i32 ^ 9i32;
   |             ^^^^^^^^^^^ help: did you mean to write: `2_i32.pow(9_i32)`

error: '^' is not the exponentiation operator
  --> $DIR/suspicious_xor.rs:15:13
   |
LL |     let _ = 2i32 ^ 2i32;
   |             ^^^^^^^^^^^ help: did you mean to write: `2_i32.pow(2_i32)`

error: '^' is not the exponentiation operator
  --> $DIR/suspicious_xor.rs:16:13
   |
LL |     let _ = 50i32 ^ 3i32;
   |             ^^^^^^^^^^^^ help: did you mean to write: `50_i32.pow(3_i32)`

error: '^' is not the exponentiation operator
  --> $DIR/suspicious_xor.rs:17:13
   |
LL |     let _ = 5i32 ^ 8i32;
   |             ^^^^^^^^^^^ help: did you mean to write: `5_i32.pow(8_i32)`

error: '^' is not the exponentiation operator
  --> $DIR/suspicious_xor.rs:18:13
   |
LL |     let _ = 2i32 ^ 32i32;
   |             ^^^^^^^^^^^^ help: did you mean to write: `2_i32.pow(32_i32)`

error: '^' is not the exponentiation operator
  --> $DIR/suspicious_xor.rs:19:13
   |
LL |     let _ = 2i32 ^ macro_test!();
   |             ^^^^^^^^^^^^^^^^^^^^ help: did you mean to write: `2_i32.pow(13)`

error: aborting due to 7 previous errors

